[FIX] payment: compute related_partner_ids only for draft payments

aj-fuentes opened the pull request:
`related_partner_ids` is used in the form view of account.payment
https://github.com/odoo/odoo/blob/43fccd7aaa83117c95babc52d60dd1c26b28335a/addons/payment/views/account_payment_views.xml#L14
It is only needed for electronic payments to display the payment token
ids (as part of its domain)
https://github.com/odoo/odoo/blob/45f5167b956521f0a183ff1b1cc75fa1b273866c/addons/payment/models/account_payment.py#L13-L21
and for draft payments only.
https://github.com/odoo/odoo/blob/43fccd7aaa83117c95babc52d60dd1c26b28335a/addons/payment/views/account_payment_views.xml#L16

Before this change the form view will fail when partner_id has a company
that is not in the context. This is not incorrect, but it is unexpected
as `related_partner_ids` is not needed to just show the form view. A way
to reproduce the issue is to confirm a payment, then change the partner
company to a company not accessible for current user.

This causes issues during migrations: the form view for the failing
payments is displayed just fine for versions <=13.0

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/92861).

AntoineVDV commented:
Hi, even though the diff is in the `payment` module, it concerns the model `account.payment` which is managed by the RD Accounting team. I see you have done your research and it seems like your fix addresses the issue so I wouldn't mind if you merge it as-is. However, should you need a review still, I'd ping the team responsible for `account.*` models since they're the experts.

Additionally, could you please update your commit message to append "payments" since it's not clear which drafts we're talking about?

aj-fuentes commented:
Hi @AntoineVDV 

Thanks for the comments. I've updated the commit message (and PR title).

> However, should you need a review still, I'd ping the team responsible for account.* models since they're the experts.

Yes, I a review will be appreciated since this could be solved alternatively with `compute_sudo=True`. I still believe that failing for draft payments seems correct which will not happen with sudo. Moreover sudo would allow to pickup saved tokens for partners from invalid companies. As per me knowledge, in the standard flow it is forbidden anyway for new payments to use a partner from another company. 

Please feel free to request review from anybody you think should be involved. I added RD-Accounting right now.

aj-fuentes commented:
Note: a similar fix is needed (I have it ready) in 15.0 but it is _not_ a forward-port of this one.

aj-fuentes commented:
Hello @qdp-odoo 
Could you review this PR please? (Or suggest another reviewer).
Thanks in advance :)

aj-fuentes commented:
Hello @odoo/rd-accounting
Could you please have a look to this small PR?

It causes upgrade failures, including RRs. Records that can be shown in the form view in 13, cannot in 14. It seems the information fetched by this field is only needed while editing draft payments, thus it's unnecessary to compute it for displaying every form view.

william-andre commented on addons/payment/models/account_payment.py:
```diff
@@ -28,12 +28,13 @@ def _get_payment_chatter_link(self):
               | p.partner_id.commercial_partner_id
               | p.partner_id.commercial_partner_id.child_ids
-            )._origin
+            )._origin if p.state == 'draft' else False
```
Wouldn't a `search` with a domain with a clause `child_of` fix this?
I think it would even reduce the number of queries done, even though this is an explicit `search` in a loop...

william-andre replied:
Or a `sudo`

aj-fuentes replied:
> Or a sudo

That would be simpler, indeed. Originally I discussed with @KangOl adding `compute_sudo=True` to the field but it seemed more intrusive to do otherwise. Let me know what you prefer and I'll update the PR.

aj-fuentes replied:
> Wouldn't a search with a domain with a clause child_of fix this?

I could test that...

aj-fuentes replied:
Hello @william-andre 

I couldn't figure out how to do this with `search`: since `partner_id` belongs to another company we cannot get `partner_id.commercial_partner_id`.
I've pushed a version using `compute_sudo=True`.
If you believe we better use `search`, please let me know how should we do that.

Thanks again for you review!

william-andre reviewed (commented):
> Before this change the form view will fail when partner_id has a company that is not in the context. 

What does it mean for a form view to fail?

> A way to reproduce the issue is to confirm a payment, then change the partner company to a company not accessible for current user.

You should be able to see and open a document with a partner you don't have access to (you see which partner it is, but you can't read anything else than it's name (and id)

william-andre reviewed (approved):
@robodoo r+

BayarkhuuBataa commented:
using `compute_sudo=True` is failed (multi company rule access warning) when selected partner has any another company child.
